# Data Acquisition

Pulls the four datasets for the Chicago Traffic Crash Analytics project and
loads each into a dataframe. All datasets are on Socrata-based open data
portals, queried over plain HTTPS (no special SDK).

- Primary 1: Traffic Crashes - Crashes
- Primary 2: Traffic Crashes - People
- Secondary 1: NOAA hourly weather (Midway station)
- Secondary 2: Chicago Traffic Tracker - Historical Congestion Estimates by Segment


In [1]:
import requests
import pandas as pd

pd.set_option("display.max_columns", 50)


## Primary 1: Traffic Crashes - Crashes

In [2]:
CRASHES_URL = "https://data.cityofchicago.org/resource/85ca-t3if.json"

# Scope to a recent 2-year window to keep the pull manageable; full history
# is available from the same endpoint by removing/adjusting $where.
params = {
    "$limit": 50000,
    "$where": "crash_date >= '2023-01-01T00:00:00' AND crash_date < '2025-01-01T00:00:00'",
}

resp = requests.get(CRASHES_URL, params=params)
resp.raise_for_status()
crashes = pd.DataFrame(resp.json())
crashes.shape


(50000, 49)

In [3]:
crashes.dtypes

crash_record_id                     str
crash_date_est_i                    str
crash_date                          str
posted_speed_limit                  str
traffic_control_device              str
device_condition                    str
weather_condition                   str
lighting_condition                  str
first_crash_type                    str
trafficway_type                     str
alignment                           str
roadway_surface_cond                str
road_defect                         str
report_type                         str
crash_type                          str
hit_and_run_i                       str
damage                              str
date_police_notified                str
prim_contributory_cause             str
sec_contributory_cause              str
street_no                           str
street_direction                    str
street_name                         str
beat_of_occurrence                  str
num_units                           str


In [4]:
crashes.head()

,crash_record_id,crash_date_est_i,crash_date,posted_speed_limit,traffic_control_device,device_condition,weather_condition,lighting_condition,first_crash_type,trafficway_type,alignment,roadway_surface_cond,road_defect,report_type,crash_type,hit_and_run_i,damage,date_police_notified,prim_contributory_cause,sec_contributory_cause,street_no,street_direction,street_name,beat_of_occurrence,num_units,crash_month,most_severe_injury,injuries_total,injuries_fatal,injuries_incapacitating,injuries_non_incapacitating,injuries_reported_not_evident,injuries_no_indication,injuries_unknown,crash_hour,crash_day_of_week,idot_control_no,latitude,longitude,location,intersection_related_i,statements_taken_i,photos_taken_i,private_property_i,work_zone_i,work_zone_type,workers_present_i,dooring_i,lane_cnt
0,0e671ca4c1c802a337a6dd0334c6cbcbdf58f09d17d8f4...,Y,2023-01-01T00:00:00.000,30,NO CONTROLS,NO CONTROLS,UNKNOWN,UNKNOWN,PARKED MOTOR VEHICLE,ONE-WAY,STRAIGHT AND LEVEL,UNKNOWN,UNKNOWN,NOT ON SCENE (DESK REPORT),NO INJURY / DRIVE AWAY,Y,"OVER $1,500",2023-01-03T10:20:00.000,UNABLE TO DETERMINE,UNABLE TO DETERMINE,8202,S,MARYLAND AVE,631,2,1,NO INDICATION OF INJURY,0,0,0,0,0,1,0,0,1,X002883193,41.7456613147,-87.6039549532,"{'type': 'Point', 'coordinates': [-87.60395495...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,258df4dedebe0503d6d4a7eda1f5e7217fea3ba2a9ffc2...,NaN,2023-01-01T00:00:00.000,30,TRAFFIC SIGNAL,FUNCTIONING PROPERLY,RAIN,"DARKNESS, LIGHTED ROAD",ANGLE,FOUR WAY,STRAIGHT AND LEVEL,WET,NO DEFECTS,NOT ON SCENE (DESK REPORT),NO INJURY / DRIVE AWAY,NaN,"OVER $1,500",2023-01-03T13:06:00.000,NOT APPLICABLE,NOT APPLICABLE,3608,W,47TH ST,821,2,1,NO INDICATION OF INJURY,0,0,0,0,0,2,0,0,1,X002883340,41.8080977724,-87.7142913571,"{'type': 'Point', 'coordinates': [-87.71429135...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,aaf05cd611c13d565f754f8b86cec048cb2a23f6375baa...,Y,2023-01-01T00:00:00.000,30,NO CONTROLS,NO CONTROLS,CLEAR,"DARKNESS, LIGHTED ROAD",PARKED MOTOR VEHICLE,ONE-WAY,STRAIGHT AND LEVEL,DRY,NO DEFECTS,NOT ON SCENE (DESK REPORT),NO INJURY / DRIVE AWAY,Y,"OVER $1,500",2023-01-02T09:45:00.000,UNABLE TO DETERMINE,UNABLE TO DETERMINE,4821,W,OHIO ST,1532,3,1,NO INDICATION OF INJURY,0,0,0,0,0,1,0,0,1,X002881786,41.8912628324,-87.7466671488,"{'type': 'Point', 'coordinates': [-87.74666714...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,e7f95d5faef6b11ea2f58813fd55e961090c65a498b354...,NaN,2023-01-01T00:00:00.000,45,NO CONTROLS,NO CONTROLS,CLEAR,"DARKNESS, LIGHTED ROAD",REAR END,DIVIDED - W/MEDIAN BARRIER,STRAIGHT AND LEVEL,DRY,NO DEFECTS,NOT ON SCENE (DESK REPORT),NO INJURY / DRIVE AWAY,NaN,"OVER $1,500",2023-01-01T02:10:00.000,UNABLE TO DETERMINE,NOT APPLICABLE,932,N,LAKE SHORE DR,1833,2,1,NO INDICATION OF INJURY,0,0,0,0,0,7,0,0,1,X002880778,41.9000816154,-87.6194148453,"{'type': 'Point', 'coordinates': [-87.61941484...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,a782165380f5c9f751daa1bb820937b0826e4f998d36ac...,Y,2023-01-01T00:01:00.000,30,NO CONTROLS,NO CONTROLS,UNKNOWN,"DARKNESS, LIGHTED ROAD",PARKED MOTOR VEHICLE,NOT DIVIDED,STRAIGHT AND LEVEL,UNKNOWN,NO DEFECTS,ON SCENE,INJURY AND / OR TOW DUE TO CRASH,Y,"OVER $1,500",2023-01-01T12:59:00.000,UNABLE TO DETERMINE,UNABLE TO DETERMINE,1140,W,MONROE ST,1232,2,1,NO INDICATION OF INJURY,0,0,0,0,0,1,0,0,1,X002882860,41.8804524162,-87.6562990571,"{'type': 'Point', 'coordinates': [-87.65629905...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


## Primary 2: Traffic Crashes - People

In [5]:
PEOPLE_URL = "https://data.cityofchicago.org/resource/u6pd-qa9d.json"

params = {
    "$limit": 50000,
    "$where": "crash_date >= '2023-01-01T00:00:00' AND crash_date < '2025-01-01T00:00:00'",
}

resp = requests.get(PEOPLE_URL, params=params)
resp.raise_for_status()
people = pd.DataFrame(resp.json())
people.shape


(50000, 28)

In [6]:
people.dtypes

person_id                str
person_type              str
crash_record_id          str
vehicle_id               str
crash_date               str
city                     str
state                    str
zipcode                  str
sex                      str
age                      str
drivers_license_state    str
drivers_license_class    str
safety_equipment         str
airbag_deployed          str
ejection                 str
injury_classification    str
driver_action            str
driver_vision            str
physical_condition       str
bac_result               str
seat_no                  str
hospital                 str
pedpedal_action          str
pedpedal_visibility      str
pedpedal_location        str
ems_agency               str
ems_run_no               str
bac_result_value         str
dtype: object

In [7]:
people.head()

,person_id,person_type,crash_record_id,vehicle_id,crash_date,city,state,zipcode,sex,age,drivers_license_state,drivers_license_class,safety_equipment,airbag_deployed,ejection,injury_classification,driver_action,driver_vision,physical_condition,bac_result,seat_no,hospital,pedpedal_action,pedpedal_visibility,pedpedal_location,ems_agency,ems_run_no,bac_result_value
0,O1979156,DRIVER,8a82d14f6d2d392638a8c5f5bdaee89ce8c42005c4529c...,1885773,2024-12-31T23:47:00.000,CHICAGO,IL,60615,F,24,IL,D,USAGE UNKNOWN,DEPLOYMENT UNKNOWN,NONE,NO INDICATION OF INJURY,FAILED TO YIELD,UNKNOWN,NORMAL,TEST NOT OFFERED,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,O1979157,DRIVER,8a82d14f6d2d392638a8c5f5bdaee89ce8c42005c4529c...,1885775,2024-12-31T23:47:00.000,CHICAGO,IL,60617,M,25,IL,D,NONE PRESENT,NOT APPLICABLE,NONE,NO INDICATION OF INJURY,NONE,UNKNOWN,NORMAL,TEST NOT OFFERED,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,P437508,PASSENGER,8a82d14f6d2d392638a8c5f5bdaee89ce8c42005c4529c...,1885775,2024-12-31T23:47:00.000,CHICAGO,IL,NaN,F,19,NaN,NaN,USAGE UNKNOWN,DID NOT DEPLOY,NONE,NO INDICATION OF INJURY,NaN,NaN,NaN,NaN,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,O1979583,DRIVER,cc89da2a2705cf16fbe17bafd4205fad11bc3853956f41...,1886186,2024-12-31T23:41:00.000,NaN,NaN,NaN,M,NaN,NaN,NaN,USAGE UNKNOWN,DID NOT DEPLOY,NONE,NO INDICATION OF INJURY,IMPROPER TURN,UNKNOWN,UNKNOWN,TEST NOT OFFERED,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,O1979167,DRIVER,3b32c74ced97162dcb27f4cab0e9b3beb389e6548608b7...,1885780,2024-12-31T23:30:00.000,INDIANAPOLIS,IN,46214,F,29,IN,NaN,USAGE UNKNOWN,NOT APPLICABLE,NONE,NO INDICATION OF INJURY,UNKNOWN,NOT OBSCURED,UNKNOWN,TEST NOT OFFERED,NaN,REFUSED-NO INJURY,NaN,NaN,NaN,NaN,NaN,NaN


## Secondary 1: NOAA hourly weather (Midway station)

Requires a free NOAA CDO API token: https://www.ncdc.noaa.gov/cdo-web/token
Set it as an environment variable `NOAA_TOKEN` before running this cell.


In [8]:
import os

NOAA_TOKEN = os.environ.get("NOAA_TOKEN")
NOAA_URL = "https://www.ncei.noaa.gov/cdo-web/api/v2/data"

# GHCND station for Midway Airport, Chicago: GHCND:USW00014819
params = {
    "datasetid": "GHCND",
    "stationid": "GHCND:USW00014819",
    "startdate": "2023-01-01",
    "enddate": "2023-01-31",
    "limit": 1000,
    "units": "standard",
}

if NOAA_TOKEN:
    resp = requests.get(NOAA_URL, headers={"token": NOAA_TOKEN}, params=params)
    resp.raise_for_status()
    weather = pd.DataFrame(resp.json().get("results", []))
else:
    weather = pd.DataFrame()
    print("NOAA_TOKEN not set, skipping live pull. See markdown above for how to get one.")

weather.shape


(317, 5)

## Secondary 2: Chicago Traffic Tracker - Historical Congestion Estimates by Segment

In [9]:
CONGESTION_URL = "https://data.cityofchicago.org/resource/4g9f-3jbs.json"

params = {
    "$limit": 50000,
    "$where": "time >= '2024-06-11T00:00:00' AND time < '2024-07-11T00:00:00'",
}

resp = requests.get(CONGESTION_URL, params=params)
resp.raise_for_status()
congestion = pd.DataFrame(resp.json())
congestion.shape


(50000, 22)

In [10]:
congestion.dtypes

time                  str
segment_id            str
speed                 str
street                str
direction             str
from_street           str
to_street             str
length                str
street_heading        str
bus_count             str
message_count         str
hour                  str
day_of_week           str
month                 str
record_id             str
start_latitude        str
start_longitude       str
end_latitude          str
end_longitude         str
start_location     object
end_location       object
comments              str
dtype: object

In [11]:
congestion.head()

,time,segment_id,speed,street,direction,from_street,to_street,length,street_heading,bus_count,message_count,hour,day_of_week,month,record_id,start_latitude,start_longitude,end_latitude,end_longitude,start_location,end_location,comments
0,2024-06-11T13:01:03.000,127,19,Ashland,NB,Addison,Irving Park,0.5,N,2,14,13,3,6,0127-202406111801,41.9469890515,-87.6686743455,41.9542892901,-87.6688650541,"{'type': 'Point', 'coordinates': [-87.66867434...","{'type': 'Point', 'coordinates': [-87.66886505...",NaN
1,2024-06-11T13:01:03.000,129,26,Ashland,NB,Montrose,Lawrence,0.5,N,1,1,13,3,6,0129-202406111801,41.9615905019,-87.6690743088,41.9688759177,-87.6692552405,"{'type': 'Point', 'coordinates': [-87.66907430...","{'type': 'Point', 'coordinates': [-87.66925524...",NaN
2,2024-06-11T13:01:03.000,131,25,Ashland,NB,Foster,Bryn Mawr,0.5,N,1,8,13,3,6,0131-202406111801,41.9761710709,-87.6694299012,41.9834747775,-87.6696322417,"{'type': 'Point', 'coordinates': [-87.66942990...","{'type': 'Point', 'coordinates': [-87.66963224...",NaN
3,2024-06-11T13:01:03.000,132,14,Ashland,NB,Bryn Mawr,Clark,0.23,N,2,10,13,3,6,0132-202406111801,41.9834747775,-87.6696322417,41.9867518859,-87.6695265292,"{'type': 'Point', 'coordinates': [-87.66963224...","{'type': 'Point', 'coordinates': [-87.66952652...",NaN
4,2024-06-11T13:01:03.000,133,26,Ashland,SB,91st,95th,0.5,S,3,21,13,3,6,0133-202406111801,41.7284832358,-87.6630347385,41.7212162434,-87.6628665545,"{'type': 'Point', 'coordinates': [-87.66303473...","{'type': 'Point', 'coordinates': [-87.66286655...",NaN


## Summary

| Dataset | Rows | Columns | Time span pulled | Geography |
|---|---|---|---|---|
| Traffic Crashes - Crashes | 50,000 | 49 | 2023-01-01 to 2024-12-31 (capped at `$limit=50000`, chronological from start of window) | Chicago |
| Traffic Crashes - People | 50,000 | 28 | 2023-01-01 to 2024-12-31 (capped at `$limit=50000`) | Chicago |
| NOAA daily weather (GHCND) | 317 | 5 | 2023-01-01 to 2023-01-31 | Midway station (GHCND:USW00014819) |
| Traffic Tracker Congestion | 50,000 | 22 | 2024-06-11 to 2024-07-11 (capped at `$limit=50000`) | Chicago arterial segments |

Note: the crashes/people/congestion pulls are capped by `$limit=50000` and are
not the full history of each dataset (crashes alone has 950K+ rows since
2017); the `$where` window can be widened later once we scope the actual
modeling dataset.
